File: Find First and Last Position of Element in Sorted Array/binarySearch.py
from typing import List

class Solution:
    def binarySearch(self, nums: List[int], target: int, findFirst: bool) -> int:
        left, right = 0, len(nums) - 1
        result = -1

        while left <= right:
            mid = (left + right) // 2

            if nums[mid] < target:
                left = mid + 1
            elif nums[mid] > target:
                right = mid - 1
            else:
                result = mid
                if findFirst:
                    right = mid - 1  # continue buscando à esquerda
                else:
                    left = mid + 1   # continue buscando à direita
        return result

    def searchRange(self, nums: List[int], target: int) -> List[int]:
        first = self.binarySearch(nums, target, True)
        last = self.binarySearch(nums, target, False)
        return [first, last]

File: Find First and Last Position of Element in Sorted Array/binarySearch.cpp
class Solution {
public:
    // Função auxiliar para encontrar a primeira ou última ocorrência
    int binarySearch(vector<int>& nums, int target, bool findFirst) {
        int left = 0, right = nums.size() - 1;
        int result = -1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (nums[mid] < target) {
                left = mid + 1;
            } else if (nums[mid] > target) {
                right = mid - 1;
            } else {
                result = mid;
                if (findFirst) {
                    right = mid - 1;  
                } else {
                    left = mid + 1; 
                }
            }
        }

        return result;
    }

    vector<int> searchRange(vector<int>& nums, int target) {
        int first = binarySearch(nums, target, true);   // busca primeira posição
        int last = binarySearch(nums, target, false);   // busca última posição
        return {first, last};
    }
};
